git_history_of_project/git_history_of_project.md

```diff
@@ -0,0 +1,503 @@
+</div>
+
+### Using the commit number
+
+If there is a particular commit you want to look back on and are not sure how far back it was, having to refer to it as `HEAD~n` could get extremely frustrating, especially since n will continue to change as you work more on your project and make more commits. Luckily there is another way to refer to a particular commit, using its **commit number**.
+
+When you enter `git commit -m "short descriptive message"` no output appears, but Git gives that commit a unique identifier, its commit number. By typing `git log` we can see ALL of the previous commits. Since that might be a lot of output to deal with, we will just ask for the last 3 commits using `git log -n 3`. The number after `-n` is how many commits you want displayed from the log.
+
+Let's see what that output looks like for Dracula:
+
+```
+$ git log -n 3
+commit 024af1d5d311c5d1a737cca3fb13d1b62c2a29ed (HEAD -> main, origin/main)
+Author: Vlad Dracula <vlad@tran.sylvan.ia>
+Date:   Wed Apr 6 13:17:26 2022 -0400
+
+    Ignore data files and the results folder
+
+commit 656471fa1f38c8b0b9a6a2fafdf3ba68610e61fa
+Author: Vlad Dracula <vlad@tran.sylvan.ia>
+Date:   Wed Apr 6 13:15:24 2022 -0400
+
+    Start files on Jupiter and Mercury
+
+commit 1e587d25f619aa0aa10fce19b44e4e71503fa41e
+Author: Vlad Dracula <vlad@tran.sylvan.ia>
+Date:   Wed Apr 6 13:13:21 2022 -0400
+
+    Add concerns about effects of Mars' moons on Wolfman
+```
+
+Use the arrow keys to scroll down if this output is longer than the window you are working in. If that is the case, you can always quit the output by pressing `Q`. Entering `git log` without the `-n` flag will give you the list of every commit, so if a project has been in progress for a while, this can be a lot of output!
+
+The first thing to notice is that your commit messages are here! This is a good reminder to write clear and concise messages because **future** you may be very grateful when trying to figure out where exactly **past** you introduced a particular issue.
+
+The second thing to notice is the structure of each entry in the log: commit, Author, Date, message.
+When you identify which commit you want to look at, the commit number is the 40 digit string of letters and numbers at the top of it after the word "commit". In Dracula's repository, the identifier for the commit in which he "add[ed] concerns about effects of Mars' moons on Wolfman" is `1e587d25f619aa0aa10fce19b44e4e71503fa41e`.
+
+<div class = "learnmore">
+**What is this number?**
+
+We are used to seeing numbers in base 10, which are made up of the ten digits $0,1,2,3,4,5,6,7,8,$ and $9$. Commit numbers are in base 16, called [hexadecimal](https://en.wikipedia.org/wiki/Hexadecimal). In this number system, we use the 10 familiar digits as well as the digits $a,b,c,d,e,$ and $f$ corresponding to 10,11,12,13,14, and 15 respectively in base 10.
```

Comment:
You have 0-9 and a-f in Latex but the rest of the numbers not... this kind of distracts me but I'm curious about what you're trying to convey, maybe I'm missing it?  I tend to like sans serif so maybe this is just a question of taste.

Reply:
This is prompting some deep exploration into my LaTeX-brain. I do think that the letters a-f need to be distinguished from regular print in some way to emphasize that they are not being used as letters but as symbols for digits. Perhaps `a,b,c,d,e,` and `f` looks less jarring?

I have in the past been a very strong believer that all math needs to be in math mode, similarly all code should be in the right font, etc. If I need to let this go, you can tell me. I already let an $n$ go in another module so it can be done!

```suggestion
We are used to seeing numbers in base 10, which are made up of ten digits. Commit numbers are in base 16, called [hexadecimal](https://en.wikipedia.org/wiki/Hexadecimal). In this number system, we use the 10 familiar digits `1,2,3,4,5,6,7,8,9` as well as the digits `a,b,c,d,e,` and `f` corresponding to 10,11,12,13,14, and 15 respectively in base 10.
```